functions/src/index.ts

```diff
@@ -46,163 +45,194 @@ const genericErrorHandler = (err: any, req: express.Request, res: express.Respon
+    return getClaimsFromJWT(token);
+  }
+};
 
-  let token: string | undefined = undefined;
-  if (req.headers.authorization && req.headers.authorization.split(' ')[0] === 'Bearer') {
-    token = req.headers.authorization.split(' ')[1];
-  } else if (req.query && req.query.token) {
-    token = req.query.token as string;
-  } else if (req.cookies && req.cookies.token) {
-    token = req.cookies.token;
-  }
-
-  getValidatedConfig()
-    .then((config) => {
-      if (!token) {
-        res.error(400, "Missing token in headers, query or cookie");
-      }
-      else if (token === "test") {
-        req.claims = {
-          user_id: "http://example.com/users/test",
-          platform_user_id: "test",
-          platform_id: "http://example.com",
-          context_id: "testContextId"
-        };
-        next();
-      }
-      else {
-        const public_key = config.admin.public_key.split("\\n").join("\n");
-        verify(token, public_key, { algorithms: ["RS256"] }, (error: any, decoded: any) => {
-          if (error) {
-            res.error(403, error);
-          }
-          else {
-            const claims = decoded.claims as JWTClaims;
-            if (!claims.user_id) {
-              res.error(403, "Missing user_id in JWT claims!");
-            }
-            if (!claims.platform_user_id) {
-              res.error(403, "Missing platform_user_id in JWT claims!");
-            }
-            if (!claims.platform_id) {
-              res.error(403, "Missing platform_id in JWT claims!");
-            }
-            // context_id is optional so don't test for it
-            // TODO: context_id is still "class_hash" rigse jwt_controller.rb - should it be class_hash here?
-            req.claims = claims;
-            next();
-          }
-        })
-      }
-    })
-    .catch(error => res.error(403, error))
+const authUsingJWTOrReadWriteToken = (req: express.Request): JWTClaims | string => {
+  const token = getToken(req);
+  if (!token) {
+    throw new Error("Missing token in headers, query or cookie");
+  } else if (token.startsWith(ReadWriteTokenPrefix)) {
+    return token;
+  } else {
+    return getClaimsFromJWT(token);
+  }
 };
 
 app.use(customHandlers);
 app.use(genericErrorHandler);
 app.use(bodyParser.json());
 app.use(bodyParser.urlencoded({ extended: false }));
 app.use(cors());
-app.use(checkAuth);
-
-const checkEnv = (req: express.Request, res: express.Response, next: express.NextFunction) => {
-  const { env } = req.query;
-  if (!env) {
-    res.error(403, "Missing env query parameter!");
-  }
-  else {
-    req.env = env as string;
-    next();
-  }
-};
 app.use(checkEnv);
 
 app.get('/api/v1/resources', (req, res) => {
-  return BaseResourceObject.FindAll(db, req.env, req.claims, req.query)
-    .then(resources => resources.map((resource) => resource.apiResult()))
-    .then(apiResults => res.success(apiResults))
-    .catch(error => res.error(400, error))
-})
+  try {
+    const claims = authUsingJWT(req);
```

Comment:
You could also do these auth checks as a single express middleware function (like `checkEnv` above) that authed based on the path and then added the `claims` or `claimsOrReadWriteToken` to the `req` object.  I'm not sure which would be cleaner so I guess leave it as is.

Reply:
I thought about that and even implemented something similar. We can also add middleware to particular handlers like:

`app.get('/api/v1/resources', authauthUsingReadWriteTokenOrJWT, (req, res) => {`
`app.get('/api/v1/resources', authUsingJWT, (req, res) => {`
`app.get('/api/v1/resources', authUsingReadWriteToken (req, res) => {`
etc.

The main problem is extending `request` object that gets messy long-term. Especially when we have properties that are optional and exclusive in fact: `claims` or `readWriteToken`. If we use this kind of middleware that sometimes assigns claims, sometimes read-write token, there's a problem with useful typings.

- We can mark these properties optional in the `Request` interface and we can explicitly check if they are present. Kinda safe, but then using middleware with error handling doesn't make sense, as we'll duplicate this logic (error handling). If auth rules changes, you need to update the code in two places. If you forget about handler, TS won't help, and users will see auth errors.
- We mark these properties optional and manually cast them in the handlers. We don't duplicate the error handling code. But typecasting (or `!` operator) is based only on the developer's knowledge. And if the implementation of auth middleware changes, types won't help (as we did casting or used `!`). I think it might be error-prone and doesn't seem clean.
- We mark them as always present in the `Request` interface. It's the same problem as above, we just don't have to cast things. But if auth changes for given method, TS won't detect an error.

So, explicit implementation in handlers seems the safest to me. Also, cleanest as we don't need additional `if`s, casting, or silent assumptions about things available in `req`.

Reply:
Another thing to consider (and perhaps you already did), is to treat the claims more generically. So they aren't just JWT claims but they are authorization claims.  Then when there is a readWrite token, it is added a claims object with a claim like `{readWriteToken: [token]}`.  Then all the code can continue to just handle claims instead of needing to include claims or a string. So this would push all of the claim checking logic down into the BaseResourceObject

Reply:
So there are two things here I think.

Yes, I was trying to implement `AuthClaims = JWTClaims | ReadWriteTokenClaims` 😉 
 ReadWriteTokenClaims was exactly `{readWriteToken: string}`. Surprisingly, it isn't much easier to work later with that as interfaces are not present in JS runtime ofc and I'd need to test for properties:
<img width="817" alt="Screenshot 2020-04-15 at 12 32 15" src="https://user-images.githubusercontent.com/767857/79328073-c7b83380-7f15-11ea-9132-106db12439f5.png">
<img width="545" alt="Screenshot 2020-04-15 at 13 12 44" src="https://user-images.githubusercontent.com/767857/79331040-d6551980-7f1a-11ea-8306-3b60f84648ec.png">

With string:
<img width="503" alt="Screenshot 2020-04-15 at 12 31 30" src="https://user-images.githubusercontent.com/767857/79328065-c4bd4300-7f15-11ea-9452-f2ac94707c37.png">

I just though the string version is a bit more bulletproof as we don't depend on property names etc. But I can change it, I don't think there's a big difference here.

> So this would push all of the claim checking logic down into the BaseResourceObject

This is in fact separate from string vs claims. We could have this generic auth middleware that generates AuthClaims or JWTClaims | string. But then it's not that clear which authentication method is required for given action. It'd be hidden in BaseResourceObject internals (or S3ResourceObject etc.). In the current implementation, we can see that more clearly I think, as handlers explicitly call authUsingJWT or authUsingJWTOrReadWriteToken. But again, that's matter of taste, I can change it if I'm missing something.

Reply:
I ended up refactoring to AuthClaims approach, it's all pretty similar, but less wordy and it would be easier to add more methods in the future if we need to.